### PR TITLE
feat(auth): Add support for signing out user from single device.

### PR DESCRIPTION
**Pull request by klkucaj**
#### Overview

This pull request introduces enhanced session management functionality, allowing users to sign out from either the current device or all devices, as well as improving session tracking with the `SessionManager`. Additionally, it includes the setup for a new Flutter test project, along with Docker configuration and GitHub Actions integration for running these tests.

#### Key Changes:
1. **New Sign-Out Methods**:
   - `signOutCurrentDevice`: Signs out the user from the current session/device only.
   - `signOutAllDevices`: Signs out the user from all active sessions on all devices.

2. **Deprecation of the `signOut` Method**:
   - The `signOut` method, which logs out from all devices, is now deprecated but still available for backward compatibility. Users should migrate to `signOutCurrentDevice` or `signOutAllDevices`.

3. **Addition of New Flutter Test Project**:
   - A new `serverpod_test_flutter` project has been added in the `tests` directory, providing an environment for running integration tests for Flutter.

4. **Docker Setup and Scripts**:
   - Docker configuration and relevant scripts (`run_tests_flutter_integration`, `start-server.sh`, etc.) have been added for running the Flutter test project in a containerized environment.

5. **GitHub Actions Integration**:
   - A new `integration_test_flutter` step has been added to the GitHub Actions workflow, allowing automated testing of the Flutter integration on different versions of Flutter (`3.19.0` and `3.24.0`).
   - 

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

### CODE REVIEW

File: modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart [hunk 1]
```diff
@@ -16,9 +16,17 @@ class StatusEndpoint extends Endpoint {
...
 
-  /// Signs out a user.
-  Future<void> signOut(Session session) async {
-    await UserAuthentication.signOutUser(session);
+  /// Signs out a user from the server. If [allSessions] is true, the user is signed out
+  /// from all active sessions by deleting all authentication keys. If [allSessions]
+  /// is false, only the current session's authentication key is deleted.
+  Future<void> signOut(
+    Session session, {
+    required bool allSessions,
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart:24]: Breaking: How will this affect already existing clients? If I understand this change correctly, all requests from old clients will fail now since this is a required parameter in the endpoint.

In the specification of the issue this was listed as additional requirements:

> This change needs to be done in a backwards-compatible way in the network layer so that old clients are still supported with preserved behavior.
> 
> This could be a configurable option so that developers can choose to migrate old clients over to signing out only a single device.

Please ensure this is respected in the change.

**Review by SandPod — CHANGES_REQUESTED** [review]
As far as I can tell this change does not follow the requirements of the issue. Please read the issue carefully again.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart (hunk 1 above)
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart]: @SandPod The issue has been addressed by refactoring the code (the code above) and replacing it with new methods. While some parts of the original code may still be present, it has largely been updated. The changes also ensure backward compatibility, so old clients should still function as expected without requiring immediate changes on their end.

File: tests/serverpod_test_server/test_e2e/signout/auth_signout_test.dart [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import 'package:serverpod_test_server/test_util/config.dart';
+import 'package:test/expect.dart';
+import 'package:test/scaffolding.dart';
+
+void main() {
+  group('Given an authenticated user', () {
+    late Client client;
+    late Client secondClient;
+
+    setUpAll(() async {
```
Comment (SandPod) [inline tests/serverpod_test_server/test_e2e/signout/auth_signout_test.dart:12]: Question: We are not validating any status endpoints in this file. Do we really need to to the method validation through the client? Couldn't we just do this as integration tests on the server?

File: tests/serverpod_test_server/test_e2e/signout/auth_signout_test.dart [hunk 3]
```diff
@@ -0,0 +1,154 @@
+import 'package:serverpod_test_client/serverpod_test_client.dart';
+import 'package:serverpod_test_server/test_util/test_key_manager.dart';
+import 'package:serverpod_test_server/test_util/config.dart';
+import 'package:test/expect.dart';
+import 'package:test/scaffolding.dart';
+
+void main() {
+  group('Given an authenticated user', () {
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tests/serverpod_test_server/test_e2e/signout/auth_signout_test.dart:8]: Before fixing anything in this file, look at the question further down in this file and in `auth_status_signout_test.dart`. These tests might be subject to be moved to integration tests on the server.

But the structural advice here should be used for any of the other tests.

File: tests/serverpod_test_server/test_e2e/signout/auth_status_signout_test.dart [hunk 4]
```diff
@@ -0,0 +1,155 @@
+import 'package:serverpod_test_client/serverpod_test_client.dart';
+import 'package:serverpod_test_server/test_util/test_key_manager.dart';
+import 'package:serverpod_test_server/test_util/config.dart';
+import 'package:test/expect.dart';
+import 'package:test/scaffolding.dart';
+
+void main() {
```
Comment (SandPod) [inline tests/serverpod_test_server/test_e2e/signout/auth_status_signout_test.dart:7]: Question: What is the difference between the tests in this file and the tests in `auth_signout_test.dart`?
Should the other file maybe be removed?

File: tests/serverpod_test_server/test_e2e/signout/auth_signout_test.dart (hunk 2 above)
Comment (SandPod) [reply tests/serverpod_test_server/test_e2e/signout/auth_signout_test.dart:12]: Noticed now that `auth_status_signout_test.dart` is basically the same file. I think that one has more value than this file if there is no difference between them.

File: tests/serverpod_test_flutter/test_integration/auth_session_manager_signout_test.dart [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+      // Verify both sessions are signed out
+      firstSessionSignedIn = await client.modules.auth.status.isSignedIn();
+      secondSessionSignedIn =
+          await secondClient.modules.auth.status.isSignedIn();
+      expect(firstSessionSignedIn, isFalse);
+      expect(secondSessionSignedIn, isFalse);
+    });
+
+    test(
+        'when calling signOutCurrentDevice from the status endpoint then only the current device session is signed out and other devices remain signed in',
```
Comment (SandPod) [inline tests/serverpod_test_flutter/test_integration/auth_session_manager_signout_test.dart:126]: Suggestion: Please apply the test restructuring here that was suggested for the e2e tests.

Also make sure to update all the descriptions so they match the behavior of the test. this description mentions calling a status endpoint but in reality it is interacting with and testing the session manager.

File: .github/workflows/dart-tests.yaml [hunk 6]
```diff
@@ -298,3 +298,17 @@ jobs:
       - name: Run unit tests
         run: dart test --concurrency=1
         working-directory: ${{ matrix.path }}
+
+  integration_test_flutter:
+    name: Serverpod integration tests (flutter)
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        flutter-version: ['3.19.0', '3.24.0']
```
Comment (SandPod) [inline .github/workflows/dart-tests.yaml:307]: Question: The matrix here mentions flutter version 3.19.0 and 3.24.0 but in the dockerfile we specifically install a single version of flutter, version 3.19, is the matrix really required for this test group?

File: modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart [hunk 7]
```diff
@@ -16,11 +16,24 @@ class StatusEndpoint extends Endpoint {
...
 
-  /// Signs out a user.
-  Future<void> signOut(Session session) async {
+  /// **[Deprecated]** Signs out a user from all sessions.
+  /// Use `signOutCurrentDevice` for the current session or `signOutAllDevices` for all sessions.
+  @Deprecated(
+      'Use signOutCurrentDevice for the current session or signOutAllDevices for all sessions. This method will be removed in future releases.')
+  Future<void> signOutUser(Session session) async {
     await UserAuthentication.signOutUser(session);
   }
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart:25]: Suggestion: Revert this change.

This is a breaking change and should not be part of this PR.
This would break all old clients.

Do we need a test that calls the old endpoint to validate we don't introduce this change?

**Review by SandPod — CHANGES_REQUESTED** [review]
A couple of comments and suggestions.
There is still a breaking change in this PR.
Please remove it.

This change should also be accompanied by a documentation update that explains the new configuration options and behaviors.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart (hunk 7 above)
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart:25]: Reverted name change. 
The test is required to ensure the method's behavior remains unchanged with no functional modifications.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart [hunk 8]
```diff
@@ -189,6 +204,7 @@ class AuthConfig {
     this.allowUnsecureRandom = false,
     this.passwordHashGenerator = defaultGeneratePasswordHash,
     this.passwordHashValidator = defaultValidatePasswordHash,
+    this.signOutOption = SignOutOption.allDevices,
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:207]: Done!

File: .github/workflows/dart-tests.yaml (hunk 6 above)
Comment (klkucaj) [reply .github/workflows/dart-tests.yaml:307]: You are right, we dont need this, thanks!

File: modules/serverpod_auth/serverpod_auth_server/lib/src/generated/endpoints.dart [hunk 9]
```diff
@@ -389,14 +389,37 @@ class Endpoints extends _i1.EndpointDispatch {
...
+          name: 'signOutUser',
           params: {},
           call: (
             _i1.Session session,
             Map<String, dynamic> params,
           ) async =>
-              (endpoints['status'] as _i7.StatusEndpoint).signOut(session),
+              (endpoints['status'] as _i7.StatusEndpoint)
+                  .
+// ignore: deprecated_member_use_from_same_package
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/generated/endpoints.dart:401]: @SandPod nope, this is added automatically when u do: serverpod generate

File: tests/serverpod_test_server/test_e2e/signout/auth_status_signout_test.dart (hunk 4 above)
Comment (klkucaj) [reply tests/serverpod_test_server/test_e2e/signout/auth_status_signout_test.dart:7]: Yes, you're right

File: tests/serverpod_test_server/test_e2e/signout/auth_signout_test.dart (hunk 2 above)
Comment (klkucaj) [reply tests/serverpod_test_server/test_e2e/signout/auth_signout_test.dart:12]: Removed this.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 10]
```diff
@@ -33,22 +33,104 @@ class UserAuthentication {
...
+  /// You can control the sign-out behavior using `AuthConfig.signOutOption`.
+  ///
+  /// Example:
+  /// ```dart
+  /// auth.AuthConfig.set(auth.AuthConfig(
+  ///   signOutOption: auth.SignOutOption.currentDevice,
+  /// ));
+  /// ```
+  @Deprecated(
+      'Use signOutCurrentDevice or signOutAllDevices. This method will be removed in future releases.')
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:112]: @SandPod done https://github.com/serverpod/serverpod/issues/2885

File: tests/serverpod_test_server/lib/src/endpoints/authentication.dart [hunk 11]
```diff
@@ -66,9 +66,18 @@ class AuthenticationEndpoint extends Endpoint {
...
     await UserAuthentication.signOutUser(session);
   }
 
+  Future<void> signOutCurrentDevice(Session session) async {
+    await UserAuthentication.signOutCurrentDevice(session);
+  }
+
+  Future<void> signOutAllDevices(Session session) async {
+    await UserAuthentication.signOutAllDevices(session);
+  }
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply tests/serverpod_test_server/lib/src/endpoints/authentication.dart:79]: We had a call for this and we have sorted it out, so all good here.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart [hunk 12]
```diff
@@ -47,6 +47,15 @@ typedef PasswordHashValidator = Future<bool> Function(
...
 });
 
+/// Enum to define the sign-out options for a user.
+enum SignOutOption {
+  /// Sign out the user from all active devices.
+  allDevices,
+
+  /// Sign out the user from the current device only.
+  currentDevice,
+}
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:57]: Done!

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 13]
```diff
@@ -33,22 +33,104 @@ class UserAuthentication {
...
+      userId: userId,
+      signoutOption: AuthConfig.current.signOutOption,
+    );
+  }
+
+  /// Signs out the user from the current session.
+  static Future<void> signOutCurrentDevice(
+    Session session, {
+    int? userId,
+  }) async {
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:128]: Fixed also this.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 14]
```diff
@@ -33,22 +33,104 @@ class UserAuthentication {
       scopeNames: scopeNames,
       method: method,
     );
-
-    session.updateAuthenticated(AuthenticationInfo(userId, scopes));
     var result = await AuthKey.db.insertRow(session, authKey);
+
+    session.updateAuthenticated(
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:45]: Added a bool flag:  bool updateSession = true

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 15]
```diff
@@ -34,21 +45,73 @@ class UserAuthentication {
...
+    await session.messages.authenticationRevoked(
+      auth.userId,
+      RevokedAuthenticationAuthId(authId: '${auth.id}'),
+    );
+
+    // Clear session authentication if the signed-out user is the currently authenticated user
+    var authInfo = await session.authenticated;
+    if (auth.userId == authInfo?.userId) {
+      session.updateAuthenticated(null);
+    }
```
[the remark this answers is not in the document]
Comment (SandPod) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:115]: Same for the new method above.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 16]
```diff
@@ -34,21 +45,73 @@ class UserAuthentication {
...
 
-  /// Signs out a user from the server and deletes all authentication keys.
-  /// This means that the user will be signed out from all connected devices.
-  static Future<void> signOutUser(Session session, {int? userId}) async {
+  /// Signs out the user from all devices.
+  static Future<void> signOutUser(
+    Session session, {
+    int? userId,
+  }) async {
     userId ??= (await session.authenticated)?.userId;
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:67]: Suggestion: Please document this behavior in the methods doumentation.

The old description captured this better. I'm not sure I like this new shorter version more.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 17]
```diff
@@ -4,20 +4,31 @@ import 'package:serverpod_auth_server/module.dart';
...
 
-/// Collects methods for authenticating users.
+/// Provides methods for authenticating users.
 class UserAuthentication {
-  /// Signs in an user to the server. The user should have been authenticated
-  /// before signing them in. Send the AuthKey.id and key to the client and
-  /// use that to authenticate in future calls. In most situations you should
-  /// use one of the auth providers instead of this method.
+  /// Signs in a user and generates an authentication key.
+  /// Sends the `AuthKey.id` and `key` to the client for future authentication.
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:10]: Suggestion: This looks strange. Please read it through a couple of times and make sure it is OK.

This line is a bit funky "Signs in a user and generates an authentication key."

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 18]
```diff
@@ -4,20 +4,31 @@ import 'package:serverpod_auth_server/module.dart';
...
+  ///
+  /// Before calling this method, ensure the user has been authenticated
+  /// through appropriate methods (e.g., email, social sign-ins).
+  /// This method only generates the authentication key after successful
+  /// authentication.
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:15]: Suggestion: This is not true, we could generate an authentication key even if we fail authentication. This method has no knowledge of that.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 19]
```diff
@@ -4,20 +4,31 @@ import 'package:serverpod_auth_server/module.dart';
...
+  ///
+  /// The authenticated user will be signed into the provided session,
+  /// and their session will be updated with the user's authentication info.
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:18]: Suggestion: This is only true if the `updateSession` bool is true.

Please harmonize this.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 20]
```diff
@@ -34,21 +45,73 @@ class UserAuthentication {
...
+    if (updateSession) {
+      session.updateAuthenticated(
+        AuthenticationInfo(
+          userId,
+          scopes,
+          authId: '${result.id}',
+        ),
+      );
+    }
     return result.copyWith(key: key);
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:59]: Question: Was the copyWith still required here?

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart [hunk 21]
```diff
@@ -47,6 +47,15 @@ typedef PasswordHashValidator = Future<bool> Function(
   void Function(Object e)? onError,
 });
 
+/// Enum to define the sign-out options for a user.
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:50]: Suggestion: Change to "Enum to define the sign-out behavior for the legacy sign out endpoint".

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart [hunk 22]
```diff
@@ -47,6 +47,15 @@ typedef PasswordHashValidator = Future<bool> Function(
...
+enum SignOutOption {
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:51]: Suggestion/Discussion: Should this be renamed to `SignOutBehavior`?

File: modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart [hunk 23]
```diff
@@ -16,9 +16,53 @@ class StatusEndpoint extends Endpoint {
...
+  @Deprecated(
+      'Use signOutDevice for the current session or signOutAllDevices for all sessions. This method will be removed in future releases.')
   Future<void> signOut(Session session) async {
-    await UserAuthentication.signOutUser(session);
+    switch (AuthConfig.current.legacyUserSignOutBehavior) {
+      case SignOutOption.currentDevice:
+        var authInfo = await session.authenticated;
+        var authKeyId = int.tryParse(authInfo?.authId ?? '');
+        if (authKeyId == null) return;
+        await UserAuthentication.revokeAuthKey(
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart:29]: Suggestion: Add an empty line here for readability.

Same for the other case.

Comment (SandPod) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart:29]: Return the response from this method for readability. Same for the other case.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart [hunk 24]
```diff
@@ -16,9 +16,53 @@ class StatusEndpoint extends Endpoint {
...
+        break;
+    }
+  }
+
+  /// Signs out a user from the current session only.
+  Future<void> signOutDevice(Session session) async {
+    var authInfo = await session.authenticated;
+    var authKeyId = int.tryParse(authInfo?.authId ?? '');
+    if (authKeyId == null) return;
+    await UserAuthentication.revokeAuthKey(
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart:51]: Suggestion: add a blank line for readability after the return and return the call to this method.

Same for the other method.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart [hunk 25]
```diff
@@ -16,9 +16,53 @@ class StatusEndpoint extends Endpoint {
...
+        if (userId == null) return;
+        await UserAuthentication.signOutUser(
+          session,
+          userId: userId,
+        );
+        break;
+    }
+  }
+
+  /// Signs out a user from the current session only.
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart:46]: Suggestion: Use the word device in the description.

I'm not sure session is the correct word here, maybe "Signs out a user from a single device"?

Comment (SandPod) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart:46]: Same feedback for the other method. Session is an more of an internal thing right?

File: modules/serverpod_auth/serverpod_auth_server/test_integration/signout/signout_lagacy_option_test.dart [hunk 26]
```diff
@@ -0,0 +1,217 @@
+import 'package:serverpod/serverpod.dart';
+import 'package:serverpod_auth_server/serverpod_auth_server.dart';
+import 'package:test/test.dart';
+import '../../test/integration/test_tools/serverpod_test_tools.dart';
+
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/test_integration/signout/signout_lagacy_option_test.dart:5]: Suggestion: This file is misspelled.

File: modules/serverpod_auth/serverpod_auth_server/test_integration/signout/signout_lagacy_option_test.dart [hunk 27]
```diff
@@ -0,0 +1,217 @@
...
+void main() {
+  withServerpod(
+    'Given different legacy sign-out options',
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_server/test_integration/signout/signout_lagacy_option_test.dart:8]: Suggestion: Place the different options in different `withServerpod`.

then you will get
`Given no legacy user sign-out option defined and user signed in to multiple devices when signing out user then user is signed out of all devices`.

Break the tests into fitting groups to achive this.

File: modules/serverpod_auth/serverpod_auth_shared_flutter/lib/src/session_manager.dart [hunk 28]
```diff
@@ -83,12 +83,20 @@ class SessionManager with ChangeNotifier {
     return refreshSession();
   }
 
-  /// Signs the user out from all connected devices. Returns true if successful.
-  Future<bool> signOut() async {
+  /// Signs the user out from their devices.
+  /// If [allDevices] is true, signs out from all devices; otherwise, signs out from the current device only.
```
Comment (SandPod) [inline modules/serverpod_auth/serverpod_auth_shared_flutter/lib/src/session_manager.dart:87]: Suggestion: Add a linebreak to stay within 80 chars width.

File: tests/serverpod_test_server/test_integration/auth_module/auth_handler_test.dart [hunk 29]
```diff
@@ -0,0 +1,76 @@
+import 'package:serverpod/serverpod.dart';
+import 'package:serverpod_auth_server/serverpod_auth_server.dart';
+import 'package:serverpod_test_server/test_util/test_serverpod.dart';
+import 'package:test/test.dart';
+
+void main() async {
```
Comment (SandPod) [inline tests/serverpod_test_server/test_integration/auth_module/auth_handler_test.dart:6]: Comment: Awesome addition! 🌟

File: tests/serverpod_test_server/test_integration/auth_module/auth_signout_test.dart [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+
+      setUp(() async {
+        session = sessionBuilder.build();
+        userName = 'test';
+        email = 'test@serverpod.dev';
+        password = 'password';
+        await Emails.createUser(session, userName, email, password);
+      });
+
+      test(
```
Comment (SandPod) [inline tests/serverpod_test_server/test_integration/auth_module/auth_signout_test.dart:23]: Suggestion: Please break this up into smaller validations and have the `when` part on a group level.

Comment (SandPod) [reply tests/serverpod_test_server/test_integration/auth_module/auth_signout_test.dart:23]: Same for the other tests.

**Review by SandPod — COMMENTED** [review]
Nice work!

Getting closer, but still a couple of comments and suggestions.

File: tests/serverpod_test_server/lib/src/endpoints/authentication.dart [hunk 31]
```diff
@@ -66,6 +66,7 @@ class AuthenticationEndpoint extends Endpoint {
   }
 
   Future<void> signOut(Session session) async {
+    // ignore: deprecated_member_use
     await UserAuthentication.signOutUser(session);
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply tests/serverpod_test_server/lib/src/endpoints/authentication.dart:70]: yes, this is not deprecated anymore, i just forgot to remove the //ingore: ... part

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 32]
```diff
@@ -34,21 +45,73 @@ class UserAuthentication {
...
+    var authInfo = await session.authenticated;
+    if (userId == authInfo?.userId) {
+      session.updateAuthenticated(null);
+    }
+  }
+
+  /// Signs out the user from the current device.
+  static Future<void> revokeAuthKey(
+    Session session, {
+    required int authKeyId,
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:94]: I am not sure if we should do this, this is business layer, someone can use it to directly pass a key, and it will be confused for them if we do this, what do you think ?

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart (hunk 20 above)
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:59]: nope ur right.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart (hunk 22 above)
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:51]: @SandPod  `LegacySignOutBehaviour` ??

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart (hunk 32 above)
[the remark this answers is not in the document]
Comment (SandPod) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:94]: I understand your perspective, in this business layer all authKeyIds are ints. But if we don't take in the string that is stored in the `AuthenticationInfo` the user of the interface needs to understand what we are storing there and how it is converted. I think that implementation details would cause a lot more digging. I think it might just be simpler to do the conversion on the inside.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart (hunk 25 above)
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/status_endpoint.dart:46]: done!

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart (hunk 22 above)
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/config.dart:51]: Thanks

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart (hunk 32 above)
[the remark this answers is not in the document]
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:94]: okay, thank you

File: util/run_tests_unit [hunk 33]
```diff
@@ -16,6 +16,7 @@ declare -a projectPaths=(
     "packages/serverpod_serialization"
     "tests/serverpod_test_client"
     "tests/serverpod_test_server"
+    "tests/serverpod_test_flutter"
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply util/run_tests_unit:19]: Yes, If no test files are found, flutter test will exit with a status of 0, indicating success.

**Review by SandPod — COMMENTED** [review]
Almost there!

Just a couple of smaller comments.

File: modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart [hunk 34]
```diff
@@ -4,20 +4,23 @@ import 'package:serverpod_auth_server/module.dart';
...
 
-/// Collects methods for authenticating users.
+/// Provides methods for authenticating users.
 class UserAuthentication {
-  /// Signs in an user to the server. The user should have been authenticated
-  /// before signing them in. Send the AuthKey.id and key to the client and
-  /// use that to authenticate in future calls. In most situations you should
-  /// use one of the auth providers instead of this method.
+  /// Signs in a user to the server. The user should have been authenticated
+  /// before signing them in. Sends the `AuthKey.id` and `key` to the client and
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart:10]: @SandPod this is how it is now
https://github.com/klkucaj/serverpod/blob/aefba282b2d6b867acc3d1c32f67692f444662d2/modules/serverpod_auth/serverpod_auth_server/lib/src/business/user_authentication.dart#L9-L13

**Review by SandPod — APPROVED** [review]
LGTM! Nice work! 🚀
